examples/chip-tool/commands/discover/DiscoverCommand.cpp

```diff
@@ -20,27 +20,27 @@
 
 constexpr uint16_t kWaitDurationInSeconds = 30;
 
-CHIP_ERROR DiscoverCommand::Run(PersistentStorage & storage, NodeId localId, NodeId remoteId)
+CHIP_ERROR DiscoverCommand::Run(NodeId localId, NodeId remoteId)
 {
-    chip::Controller::CommissionerInitParams params;
+    CHIP_ERROR err;
 
-    params.storageDelegate                = &storage;
-    params.mDeviceAddressUpdateDelegate   = this;
-    params.operationalCredentialsDelegate = &mOpCredsIssuer;
+    {
+        chip::DeviceLayer::StackLock lock;
 
-    ReturnErrorOnFailure(mCommissioner.SetUdpListenPort(storage.GetListenPort()));
-    ReturnErrorOnFailure(mCommissioner.Init(localId, params));
-    ReturnErrorOnFailure(mCommissioner.ServiceEvents());
-
-    ReturnErrorOnFailure(RunCommand(mNodeId, mFabricId));
+        GetExecContext()->commissioner->RegisterDeviceAddressUpdateDelegate(this);
+        err = RunCommand(mNodeId, mFabricId);
+        SuccessOrExit(err);
+    }
 
     UpdateWaitForResponse(true);
```

Comment:
This does not look right to me.  The order here will be:

1. Take stack lock, send command, release stack lock.
2. Take `cvWaitingForResponseMutex`.
3. Set `mWaitingForResponse` to true.
4. Release ` cvWaitingForResponseMutex`.
5. Block on a condvar until `mWaitingForResponse` becomes false.

What if the response comes in between steps 1 and 2?  Won't we end up blocking on a response that will never come (because it already did) until we time out?

Or am I missing something that makes this safe?

`UpdateWaitForResponse` takes a lock, so we need to be a bit careful, but I _think_ if we are consistent about only calling `UpdateWaitForResponse` while the stack lock is held we will be OK because then the acquisition order will always be the same so we can't deadlock...

Alternately, if we do the initial `UpdateWaitForResponse` before we ever send the request, which is what `PairingCommand` does, for example.  That might be simpler.

Note that Tsan won't detect this problem because access to `mWaitingForResponse` is in fact synchronized so there is no _data_ race...

Reply:
Will fix this across the board for all command sending logic.